### PR TITLE
overwrite nullables

### PR DESCRIPTION
**Pull request by amdwit**
Here an idea for the nullables: let put them together in one .proto, generate the code and overwrite toObject, getValue and setValue. 

```
proto.NullableBoolValue.prototype.getValue = function() {
  return jspb.Message.getBooleanFieldWithDefault(msg, 1, false)
      ? null
      : jspb.Message.getBooleanFieldWithDefault(msg, 2, false);
};

proto.NullableBoolValue.prototype.setValue = function(value) {
   value!==null&& jspb.Message.setProto3BooleanField(this, 2, value)
  return jspb.Message.setProto3BooleanField(this, 1, value===null);
};

proto.NullableBoolValue.toObject = function(includeInstance, msg) {
  return jspb.Message.getBooleanFieldWithDefault(msg, 1, false)
      ? null
      : jspb.Message.getBooleanFieldWithDefault(msg, 2, false);
};
}

```

Also for the backend in Python. Could that make our work easier?

Included are the .proto . A generated and changed definition file and the generated unchanged implementation.

### CODE REVIEW

File: app/frontend/src/pb/nullables_pb.d.ts [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+  static deserializeBinary(bytes: Uint8Array): NullableBoolValue;
+  static deserializeBinaryFromReader(message: NullableBoolValue, reader: jspb.BinaryReader): NullableBoolValue;
+}
+
+export namespace NullableBoolValue {
+  export type AsObject = {
+    isNull: boolean,
+    value: boolean,
+  }
+}
```
Comment (amdwit) [inline app/frontend/src/pb/nullables_pb.d.ts:25]:
```suggestion
export namespace NullableBoolValue {
  export type AsObject = boolean | null
}
```

File: app/frontend/src/pb/nullables_pb.d.ts [hunk 2]
```diff
@@ -0,0 +1,26 @@
+import * as jspb from 'google-protobuf'
+
+
+
+export class NullableBoolValue extends jspb.Message {
+  getIsNull(): boolean;
+  setIsNull(value: boolean): NullableBoolValue;
+
+  getValue(): boolean;
```
Comment (amdwit) [inline app/frontend/src/pb/nullables_pb.d.ts:9]:
```suggestion
  getValue(): boolean | null;
```

File: app/frontend/src/pb/nullables_pb.d.ts [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+  setValue(value: boolean): NullableBoolValue;
```
Comment (amdwit) [inline app/frontend/src/pb/nullables_pb.d.ts:10]:
```suggestion
  setValue(value: boolean | null): NullableBoolValue;
```

**Comment by kalvdans** [comment]
The `google.protobuf.*Value` messages are already optional, so `google.protobuf.BoolValue` can have three values; true, false or absent. Do you want four values? (true, false, null and absent)

**Comment by amdwit** [comment]
Are you sure about that? Aren't absent values replaced by their default values on unpacking/deserializing?

**Comment by aapeliv** [comment]
> Are you sure about that? Aren't absent values replaced by their default values on unpacking/deserializing?

Yes, but messages within a message are retained, so you can tell if the message (which only contains the field) is present or not.

**Comment by amdwit** [comment]
Are we talking about the same ? Those nullables are already in the protos. I only want to add some code to js (and also Python) to make working with them easier. I only moved the nullable to its own .proto for illustrative purposes. I do not mean to change the .proto s and generated code.

**Comment by amdwit** [comment]
Also a toObject function that recursively/deep runs all toObject methods would be nice. But that will give typing challenges in typescript.

**Comment by aapeliv** [comment]
So there are two things here, the builtin `*Value` types as @kalvdans points out. These can either be absent, or have a value. E.g. in bool, you get absent/true/false.

Then there are our `Nullable*Value` that can be either absent, null, or have a value. E.g. with bool, you get absent/null/true/false.

The `Nullable*Value` types are used when a user updates their profile, so we can tell whether they didn't actually touch that field (absent), whether they want to unset it (null), or what value they want to set (yes/no). E.g. with `last_minute`, the field can have three values in the DB, null/true/false, and we allow the user to set it to any of those.

(In the other PR, #435, you should use the builtin `*Value` types if you want to return "absent" or a value. I didn't catch it in my review.)

**Comment by aapeliv** [comment]
In terms of implementation, I agree, adding some functionality to deal with these easier is a good idea. However, I don't think we should touch the generated code (or check it into git). It's a bit clunky to work around it, but it'll end up cleaner and easier, I believe, to do these modifications outside the generated chunks.

**Comment by amdwit** [comment]
The overwriting should be done by importing the generated code in another module.

**Comment by amdwit** [comment]
I suppose everybody knows about it now.

**Pull request closed without merging**